### ISSUE TITLE
vmagent: agent keeps watching with invalid config after reload

### ISSUE THREAD

**Issue by rbakhtaraev** [VictoriaMetrics/VictoriaMetrics#4850]
### Describe the bug

I have next configuration section:
```
    scrape_configs:
      - job_name: application-monitoring
        scrape_interval: 30s
        metrics_path: '/metrics'
  
        kubernetes_sd_configs:
          - role: service
            namespaces:
              names:
                - some-namespace
            selectors:
              - role: service
                label: "app: my-app"
```

With this configuration I've got next errors:
```
2023-08-17T11:37:49.311Z	error	VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:592	unexpected status code for request to "https://10.0.0.1:443/api/v1/namespaces/some-namespace/services?labelSelector=app%3Amy-app": 400; want 200; response: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"unable to parse requirement: invalid label key \\\"app:some-namespace\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"reason\":\"BadRequest\",\"code\":400}\n"
2023-08-17T11:38:19.312Z	error	VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:592	unexpected status code for request to "https://10.0.0.1:443/api/v1/namespaces/some-namespace/services?labelSelector=app%3Amy-app": 400; want 200; response: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"unable to parse requirement: invalid label key \\\"app:some-namespace\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"reason\":\"BadRequest\",\"code\":400}\n"
```

Then I've realized that I made a mistake and fixed my configuration (`:` -> `=`) :
```
    scrape_configs:
      - job_name: application-monitoring
        scrape_interval: 30s
        metrics_path: '/metrics'
  
        kubernetes_sd_configs:
          - role: service
            namespaces:
              names:
                - some-namespace
            selectors:
              - role: service
                label: "app=my-app"
```

Applied it and checked logs:

```
2023-08-17T11:38:19.312Z	error	VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:592	unexpected status code for request to "https://10.0.0.1:443/api/v1/namespaces/some-namespace/services?labelSelector=app%3Amy-app": 400; want 200; response: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"unable to parse requirement: invalid label key \\\"app:some-namespace\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"reason\":\"BadRequest\",\"code\":400}\n"

2023-08-17T11:38:28.257Z	info	VictoriaMetrics/lib/promscrape/config.go:133	restarting service discovery routines...
2023-08-17T11:38:28.257Z	info	VictoriaMetrics/lib/promscrape/config.go:176	restarted service discovery routines in 0.000 seconds, stopped=0, started=0, restarted=1
2023-08-17T11:38:28.257Z	info	VictoriaMetrics/lib/promscrape/scraper.go:201	found changes in "/etc/victoriametrics/base-configs/base.yaml"; applying these changes
2023-08-17T11:38:28.258Z	info	VictoriaMetrics/lib/promscrape/scraper.go:431	kubernetes_sd_configs: added targets: 1, removed targets: 0; total targets: 1
2023-08-17T11:38:49.315Z	error	VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:592	unexpected status code for request to "https://10.0.0.1:443/api/v1/namespaces/some-namespace/services?labelSelector=app%3Amy-app": 400; want 200; response: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"unable to parse requirement: invalid label key \\\"app:some-namespace\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"reason\":\"BadRequest\",\"code\":400}\n"


...


2023-08-17T11:57:19.390Z	error	VictoriaMetrics/lib/promscrape/discovery/kubernetes/api_watcher.go:592	unexpected status code for request to "https://10.0.0.1:443/api/v1/namespaces/some-namespace/services?labelSelector=app%3Amy-app": 400; want 200; response: "{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"unable to parse requirement: invalid label key \\\"app:some-namespace\\\": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')\",\"reason\":\"BadRequest\",\"code\":400}\n"
```

As can we see in the logs, the problem isn't fixed - vmagent continues spam this message and send incorrect requests to kube-apiserver, but correct target was added.

Also we can see this in kube-apiserver logs at the same time:
```
Couldn't parse request url.Values{"labelSelector":[]string{"app:my-app"}}: unable to parse requirement: invalid label key "app:my-app": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName', or 'my.name', or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
```

I expect that vmagent will use new configuration and won't watch "old resources".

### To Reproduce

See below

### Version

root@vmagent-test-7855c5c49f-wfrst:/# vmagent-prod --version
vmagent-20230602-100256-tags-v1.91.2-0-gde9481208

**Comment by valyala** [comment]
It has been discovered that the commit mentioned above could lead to excess churn in connections and requests to Kubernetes API server on every config reload. This has been addressed in the commit 03fece44e0f3803e16e5d9f3d71423a79b6d8785 .

**Comment by valyala** [comment]
FYI, the commit 30a645cd825e58b146c0f9c8c2640438bb139a5d fixes a few remaining issues:

- It properly updates `vm_promscrape_discovery_kubernetes_url_watchers` and `vm_promscrape_discovery_kubernetes_group_watchers` metrics.
- It properly stops an auxiliary goroutine, which is used for updating internal data structures  per each polling connection to K8S API server.
- It logs the event when the unused polling connection to k8s API server is stopped.

**Comment by valyala** [comment]
FYI, it has been appeared that the fix for this issue could trigger another issue described [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/5216#issuecomment-1782793812). The bugfix for this issue is located in the commit 632d788b63bcead6b1afa0ba8d4c2fe1a129ca92 , which will be included in the next release.

**Comment by valyala** [comment]
FYI, the issue mentioned in [this comment](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4850#issuecomment-1782796170) has been fixed in [v1.93.7 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.7).
